### ISSUE TITLE
resolve FutureWarning

### ISSUE THREAD

**Issue by mathematicalmichael** [UT-CHG/BET#320]
don't have access to add this to the projects, but this is a low-priority to-do that I am putting here for later reference.

parallel nosetests:
```
if np.issubdtype(dtype, ptype):
/home/jovyan/work/BET/bet/util.py:79: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
```

let's try to get rid of this warning, yeah?

**Comment by mathematicalmichael** [comment]
another one:
`contaminant.ipynb`
Cell 6
![Screen Shot 2019-04-28 at 11 03 34 AM](https://user-images.githubusercontent.com/40366263/56867681-52222300-69a5-11e9-9304-b6d189765995.png)

**Comment by mathematicalmichael** [comment]
found similar issues in other scientific libraries:
https://github.com/spacetelescope/stsynphot_refactor/pull/56/files
https://github.com/RaRe-Technologies/gensim/issues/2188

Can run `mpirun -n 2 nosetests test/test_sampling/test_basicSampling.py` to trigger the `FutueWarning` and then try changing `line 79` in `bet/util` to 

```
            if np.issubdtype(dtype, np.floating):
```

and it will go away, with the test still passing (oddly). 
So, that makes me wonder... why is it that if instead of changing the line in `util`, changing `line 173/175` in `Comm` to 

```
        #: float type
        self.DOUBLE = np.floating
        #: int type
        self.INT = np.integer
```

still allows the warning to spring up. Given how these parameters relate (`util` in `line 12`), the aforementioned should have done the trick.

**Comment by mathematicalmichael** [comment]
maybe some more insight: because i had `mpi4py` installed, the `MPI_for_no_mpi4py` class never got instantiated, meaning the `MPI.INT` and `MPI.FLOAT` were inhereted from the import of `MPI` in `line 180` of `Comm`. 

So, naturally I uninstall mpi4py and re-ran that test. First with the code as-it-was. Failure looked like:
```
[BET]$ mpirun -n 2 nosetests test/test_sampling/test_basicSampling.py 
.EE..........
======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.user_samples`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 696, in test_compute_QoI_and_create_discretization
    input_sample_set, savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 175, in verify_compute_QoI_and_create_discretization
    saved_disc = bet.sample.load_discretization(savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 1340, in load_discretization
    discretization_name+'_output_sample_set')
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 136, in load_sample_set
    mdat = sio.loadmat(file_name)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 217, in loadmat
    MR, _ = mat_reader_factory(f, **kwargs)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 72, in mat_reader_factory
    mjv, mnv = get_matfile_version(byte_stream)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/miobase.py", line 224, in get_matfile_version
    raise MatReadError("Mat file appears to be empty")
scipy.io.matlab.miobase.MatReadError: Mat file appears to be empty
-------------------- >> begin captured stdout << ---------------------
11t11 1
HERE HERE {'__header__': b'MATLAB 5.0 MAT-file Platform: posix, Created on: Thu Jun 13 13:47:13 2019', '__version__': '1.0', '__globals__': [], 'num_samples': array([[4]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_input_sample_set_dim': array([[1]]), 'default_input_sample_set_p_norm': array([[2.]]), 'default_input_sample_set_values': array([[1.],
       [1.],
       [1.],
       [1.]]), 'default_input_sample_set_values_local': array([[1.],
       [1.],
       [1.],
       [1.]]), 'default_input_sample_set_sample_set_type': array(['bet.sample.sample_set'], dtype='<U21'), 'default_output_sample_set_dim': array([[1]]), 'default_output_sample_set_p_norm': array([[2.]]), 'default_output_sample_set_values': array([[0.84147098],
       [0.84147098],
       [0.84147098],
       [0.84147098]]), 'default_output_sample_set_values_local': array([[0.84147098],
       [0.84147098],
       [0.84147098],
       [0.84147098]]), 'default_output_sample_set_sample_set_type': array(['bet.sample.sample_set'], dtype='<U21')} 4
1t1 1

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.create_random_discretization`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 812, in test_create_random_discretization
    savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 243, in verify_create_random_discretization
    saved_disc = bet.sample.load_discretization(savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 1348, in load_discretization
    discretization_name+attrname))
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 136, in load_sample_set
    mdat = sio.loadmat(file_name)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 217, in loadmat
    MR, _ = mat_reader_factory(f, **kwargs)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 72, in mat_reader_factory
    mjv, mnv = get_matfile_version(byte_stream)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/miobase.py", line 224, in get_matfile_version
    raise MatReadError("Mat file appears to be empty")
scipy.io.matlab.miobase.MatReadError: Mat file appears to be empty
-------------------- >> begin captured logging << --------------------
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 12 tests in 0.553s

FAILED (errors=2)
...........
----------------------------------------------------------------------
Ran 12 tests in 6.345s

OK

```
So the failures were allowed??

Now let's change to `np.floating` and `np.integer`

```
[BET]$ mpirun -n 2 nosetests test/test_sampling/test_basicSampling.py 
EE..........
======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.user_samples`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 696, in test_compute_QoI_and_create_discretization
    input_sample_set, savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 159, in verify_compute_QoI_and_create_discretization
    input_sample_set, savefile, globalize=True)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 424, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 282, in save
    globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 1197, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 217, in loadmat
    MR, _ = mat_reader_factory(f, **kwargs)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 72, in mat_reader_factory
    mjv, mnv = get_matfile_version(byte_stream)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/miobase.py", line 224, in get_matfile_version
    raise MatReadError("Mat file appears to be empty")
scipy.io.matlab.miobase.MatReadError: Mat file appears to be empty
-------------------- >> begin captured stdout << ---------------------
11t11 1

--------------------- >> end captured stdout << ----------------------

======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.create_random_discretization`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 812, in test_create_random_discretization
    savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 223, in verify_create_random_discretization
    input_domain, savefile, num_samples=num_samples, globalize=True)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 472, in create_random_discretization
    savefile, globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 424, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 282, in save
    globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 1197, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 218, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio5.py", line 292, in get_variables
    res = self.read_var_array(hdr, process)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio5.py", line 252, in read_var_array
    return self._matrix_reader.array_from_header(header, process)
  File "mio5_utils.pyx", line 673, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header
  File "mio5_utils.pyx", line 703, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header
  File "mio5_utils.pyx", line 777, in scipy.io.matlab.mio5_utils.VarReader5.read_real_complex
ValueError: cannot reshape array of size 0 into shape (1,1)

----------------------------------------------------------------------
Ran 12 tests in 0.378s

FAILED (errors=2)
E...........
======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.user_samples`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 696, in test_compute_QoI_and_create_discretization
    input_sample_set, savefile)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/test/test_sampling/test_basicSampling.py", line 159, in verify_compute_QoI_and_create_discretization
    input_sample_set, savefile, globalize=True)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 424, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sampling/basicSampling.py", line 282, in save
    globalize=globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 1197, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Dropbox/Coding/academic/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio.py", line 218, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio5.py", line 292, in get_variables
    res = self.read_var_array(hdr, process)
  File "/home/mpilosov/anaconda3/lib/python3.6/site-packages/scipy/io/matlab/mio5.py", line 252, in read_var_array
    return self._matrix_reader.array_from_header(header, process)
  File "mio5_utils.pyx", line 673, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header
  File "mio5_utils.pyx", line 703, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header
  File "mio5_utils.pyx", line 777, in scipy.io.matlab.mio5_utils.VarReader5.read_real_complex
ValueError: cannot reshape array of size 0 into shape (1,1)
-------------------- >> begin captured stdout << ---------------------
11t11 1

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 12 tests in 2.886s

FAILED (errors=1)

```


So... I'm not sure how helpful that is, but I certainly think it is related to #32 #333 #289 @lcgraham

**Comment by mathematicalmichael** [comment]
I figured out how to suppress it but I'm not sure if it is correct:

```
def get_global_values(array, shape=None):
    """
    Concatenates local arrays into global array using :meth:`np.vstack`.

    :param array: Array.
    :type P_samples: :class:`~numpy.ndarray`
    :rtype: :class:`~numpy.ndarray`
    :returns: array
    """
    if comm.size == 1:
        return array
    else:
        # Figure out the subtype of the elements of the array
        dtype = array.dtype
        mpi_dtype = False
        for ptype in possible_types.keys():
            # suppress FutureWarning 
            if ptype is int:
                ptype = np.integer
            elif ptype is float:
                ptype = np.floating

            if np.issubdtype(dtype, ptype):
                mpi_dtype = True
                dtype = ptype

        if shape is None or not mpi_dtype:
            # do a lowercase allgather
            a_shape = len(array.shape)
            array = comm.allgather(array)
            if a_shape <= 1:
                return np.hstack(array)
            else:
                return np.vstack(array)
        else:
            # do an uppercase Allgather
            whole_a = np.empty(shape, dtype=dtype)
            comm.Allgather([array.ravel(), possible_types[dtype]], [whole_a,
                                                                    possible_types[dtype]])
            return whole_a
```

**Comment by mathematicalmichael** [comment]
that led to failed parallel util tests because I changed the valid keys in the dictionary `possible_types`
I think instead we can just use 

```
            if ptype is int or ptype is MPI.INT:
                comp = np.integer
            elif ptype is float or ptype is MPI.FLOAT:
                comp = np.floating

            if np.issubdtype(dtype, comp):
                mpi_dtype = True
                dtype = ptype
```

This seems to allow that test to pass

**Comment by mathematicalmichael** [comment]
update: this change leads to just 1 error in parallel tests. but it's the same one that always gives me trouble with the HOT/COLD start mentioned in #333

**Comment by mathematicalmichael** [comment]
fixed in https://github.com/mathematicalmichael/BET/pull/22
will be merging this into my master branch after #340 is merged, submit a new PR (#345) with that change (yes, it will decrease coverage but barely). @smattis

**Comment by mathematicalmichael** [comment]
will do this after #347 is merged.
